system-addon/common/Reducers.jsm

```diff
@@ -97,16 +97,43 @@ function TopSites(prevState = INITIAL_STATE.TopSites, action) {
       return Object.assign({}, prevState, {
         editForm: {
           index: prevState.editForm.index,
-          screenshotRequestFailed: true
+          screenshotRequestFailed: true,
+          requestedScreenshotURL: null
         }
       });
     case at.SCREENSHOT_PREVIEW:
+      if (prevState.editForm &&
+        prevState.editForm.requestedScreenshotURL === action.data.customScreenshotURL) {
+        return Object.assign({}, prevState, {
+          editForm: {
+            index: prevState.editForm ? prevState.editForm.index : -1,
+            screenshotRequestFailed: false,
+            screenshotPreview: action.data.screenshotPreview,
+            customScreenshotURL: action.data.customScreenshotURL,
+            requestedScreenshotURL: null
+          }
+        });
+      }
+      return prevState;
+    case at.PREVIEW_REQUEST:
```

Comment:
Because these Reducers are also run in main, are there concerns with the main redux state updating and sending editForm state in rehydration messages?

Reply:
I don't think that should happen. `PREVIEW_REQUEST` checks `prevState.editForm` which is null for the main reducer so `editForm` should always be null. Do you think we should have separate `request` actions (one to trigger the fetch in main, another to update the local reducer) to make the separation clearer?

Reply:
Ah good point. I think this is a neat way to avoid updating main. Where TOP_SITES_EDIT to trigger reducing these actions is only seen in content.

Reply:
Ah I see the confusion. I meant "good point" referring to your first 2 sentences. I don't think there should be a separate request action for main.